### PR TITLE
Resolves #56: Add rules for planning sort-only queries as index scans.

### LINKED ISSUE #56: Add planner rules for implementing sort-only queries as index scans
Write rules for the new planner for queries with sorts, but no filters. This requires porting over the core logic in `RecordQueryPlanner.planNoFilter()` and `RecordQueryPlanner.planSortOnly()`.

### PR DESCRIPTION
**Pull request by nschiefer**
Adds a variety of rules for planning sorts as index scans. With some minor modifications, this also supports pushing down filters that can be satisfied by the same index to the index scan level using the methods in `ScanComparisons`. As a result, a sizeable number of our current planning tests now pass against the experimental planner. Also fix some bugs in the implementation of phasing in the `RewritePlanner`.

I took a major shortcut in implementing this by using the existing `KeyExpression.isPrefixKey()` method rather than writing one that can properly deal with `ExpressionRef`s that aren't `get()`-able (like groups). This is fine for now but we'll have to deal with it at some point (and it's pretty annoying).

Lastly, the `DualPlannerExtension` required some crazy hijinks to make it play nicely with the `ParameterizedTestExtension`. I know it isn't pretty but it works. :)

### CODE REVIEW

File: fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/query/plan/temp/RewritePlanner.java [hunk 1]
```diff
@@ -54,6 +54,8 @@
     @Nonnull
     private final RecordStoreState recordStoreState;
     @Nullable
+    private SingleExpressionRef<PlannerExpression> root;
```
Comment (MMcM) [inline fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/query/plan/temp/RewritePlanner.java]: Is this for future use? As near as I can tell, `root` and `currentRoot` are still always the same.

File: fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/query/plan/temp/rules/ImplementFilterRule.java [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+import com.apple.foundationdb.record.query.plan.temp.SingleExpressionRef;
+import com.apple.foundationdb.record.query.plan.temp.expressions.LogicalFilterExpression;
+import com.apple.foundationdb.record.query.plan.temp.matchers.ExpressionMatcher;
+import com.apple.foundationdb.record.query.plan.temp.matchers.ReferenceMatcher;
+import com.apple.foundationdb.record.query.plan.temp.matchers.TypeMatcher;
+
+import javax.annotation.Nonnull;
+
+/**
+ * A rule that implements a logical filter around a {@link RecordQueryPlan } as a {@link RecordQueryFilterPlan}.
```
Comment (MMcM) [inline fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/query/plan/temp/rules/ImplementFilterRule.java]: extra space (which at least this UI thinks merges the two links).

File: fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/query/plan/temp/RewritePlanner.java (hunk 1 above)
Comment (nschiefer) [reply fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/query/plan/temp/RewritePlanner.java]: No, I think it must be leftover from some refactoring that I did and then un-did. I removed it.

File: fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/query/plan/temp/rules/ImplementFilterRule.java (hunk 2 above)
Comment (nschiefer) [reply fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/query/plan/temp/rules/ImplementFilterRule.java]: Removed.

**Pull request merged**